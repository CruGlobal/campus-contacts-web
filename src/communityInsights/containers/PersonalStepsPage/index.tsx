import React, { useContext } from 'react';
import { useTranslation } from 'react-i18next';
import moment from 'moment';

import Card from '../../components/Card';
import { AppContext } from '../../../appContext';
import StagesSummary from '../../components/StagesSummary';
import StepsChart from '../../components/StepsChart';
import ImpactInfo from '../../components/ImpactInfo';
import FiltersChart from '../../components/FiltersChart';

import {
    GET_IMPACT_REPORT_MOVED,
    GET_IMPACT_REPORT_STEPS_TAKEN,
    GET_STAGES_REPORT_MEMBER_COUNT,
    GET_STEPS_COMPLETED_REPORT,
    GET_STAGES_REPORT_STEPS_ADDED,
    GET_TOTAL_STEPS_COMPLETED_SUMMARY,
} from './queries';
// GET_IMPACT_REPORT_STEPS_TAKEN INTERFACES
import { impactReportPersonalStepsCompletedCount } from './__generated__/impactReportPersonalStepsCompletedCount';
// GET_TOTAL_STEPS_COMPLETED_SUMMARY INTERFACES
import {
    communityReportStagesPersonalStepsCompleted,
    communityReportStagesPersonalStepsCompleted_community_report_stagesReport as communityReportStagesPersonalStepsCompletedEntry,
} from './__generated__/communityReportStagesPersonalStepsCompleted';
// GET_STEPS_COMPLETED_REPORT INTERFACES
import {
    communityReportDaysPersonalSteps,
    communityReportDaysPersonalSteps_community_report_daysReport as communityReportDaysPersonalRow,
    communityReportDaysPersonalSteps_community_report_daysReport_stageResults as communityReportDaysPersonalStage,
} from './__generated__/communityReportDaysPersonalSteps';
// GET_STAGES_REPORT_STEPS_ADDED INTERFACES
import {
    communityReportStagesPersonalStepsAdded,
    communityReportStagesPersonalStepsAdded_community_report_stagesReport as communityReportStagesPersonalStepsAddedRow,
} from './__generated__/communityReportStagesPersonalStepsAdded';
// GET_IMPACT_REPORT_MOVED INTERFACES
import { impactReportStageProgressionCount } from './__generated__/impactReportStageProgressionCount';
// GET_STAGES_REPORT_MEMBER_COUNT INTERFACES
import {
    communityReportStagesPersonalMemberCount,
    communityReportStagesPersonalMemberCount_community_report_stagesReport as communityReportStagesPersonalMemberCountRow,
} from './__generated__/communityReportStagesPersonalMemberCount';

const PersonalStepsPage = () => {
    const { t } = useTranslation('insights');
    const { orgId } = useContext(AppContext);

    return (
        <div>
            <ImpactInfo
                query={GET_IMPACT_REPORT_STEPS_TAKEN}
                text={(report: impactReportPersonalStepsCompletedCount) =>
                    t('personalSteps.taken', {
                        count:
                            report.community.impactReport
                                .personalStepsCompletedCount,
                    })
                }
                variables={{ id: orgId }}
            />
            <Card title={t('personalSteps.completedTotal')}>
                <StagesSummary
                    query={GET_TOTAL_STEPS_COMPLETED_SUMMARY}
                    variables={{ id: orgId }}
                    mapData={(
                        data: communityReportStagesPersonalStepsCompleted,
                    ) =>
                        data.community.report.stagesReport.map(
                            (
                                entry: communityReportStagesPersonalStepsCompletedEntry,
                            ) => ({
                                stage: entry.stage.name,
                                icon: entry.stage.name
                                    .toLowerCase()
                                    .replace(' ', '-'),
                                count: entry.personalStepsCompletedCount,
                            }),
                        )
                    }
                />
            </Card>
            <Card
                title={t('personalSteps.completed')}
                subtitle={t('personalSteps.completedSubtitle')}
            >
                <FiltersChart
                    query={GET_STEPS_COMPLETED_REPORT}
                    variables={{ id: orgId }}
                    mapData={(data: communityReportDaysPersonalSteps) =>
                        data.community.report.daysReport.map(
                            (row: communityReportDaysPersonalRow) => ({
                                ['total']: row.personalStepsCompletedCount,
                                ['stages']: row.stageResults.map(
                                    (
                                        stage: communityReportDaysPersonalStage,
                                    ) => ({
                                        name: stage.stage.name,
                                        count:
                                            stage.personalStepsCompletedCount,
                                    }),
                                ),
                                ['date']: row.date,
                            }),
                        )
                    }
                    currentDate={moment().toDate()}
                    label={t('personalSteps.legend')}
                />
            </Card>
            <Card
                title={t('personalSteps.added')}
                subtitle={t('personalSteps.addedSubtitle')}
            >
                <StepsChart
                    query={GET_STAGES_REPORT_STEPS_ADDED}
                    mapData={(data: communityReportStagesPersonalStepsAdded) =>
                        data.community.report.stagesReport.map(
                            (
                                row: communityReportStagesPersonalStepsAddedRow,
                            ) => ({
                                [t(
                                    'personalSteps.label',
                                )]: row.personalStepsAddedCount,
                                [t('stage')]: row.stage.name.toUpperCase(),
                            }),
                        )
                    }
                    label={t('personalSteps.label')}
                    index={t('stage')}
                    variables={{
                        period: 'P10Y',
                        id: orgId,
                        endDate: moment().format(),
                    }}
                />
            </Card>

            <ImpactInfo
                query={GET_IMPACT_REPORT_MOVED}
                text={(report: impactReportStageProgressionCount) =>
                    t('personalSteps.reached', {
                        count:
                            report.community.impactReport.stageProgressionCount,
                    })
                }
                variables={{ id: orgId }}
            />
            <Card
                title={t('personalSteps.members')}
                subtitle={t('personalSteps.membersSubtitle')}
            >
                <StepsChart
                    query={GET_STAGES_REPORT_MEMBER_COUNT}
                    mapData={(data: communityReportStagesPersonalMemberCount) =>
                        data.community.report.stagesReport.map(
                            (
                                row: communityReportStagesPersonalMemberCountRow,
                            ) => ({
                                [t('members')]: row.memberCount,
                                [t('stage')]: row.stage.name.toUpperCase(),
                            }),
                        )
                    }
                    label={t('members')}
                    index={t('stage')}
                    variables={{
                        period: 'P10Y',
                        id: orgId,
                        endDate: moment().format(),
                    }}
                />
            </Card>
        </div>
    );
};

export default PersonalStepsPage;
