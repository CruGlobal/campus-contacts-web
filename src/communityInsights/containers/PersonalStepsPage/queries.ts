import gql from 'graphql-tag';

const GET_IMPACT_REPORT_MOVED = gql`
    query impactReportStageProgressionCount($id: ID!) {
        community(id: $id) {
            impactReport {
                stageProgressionCount
            }
        }
    }
`;

const GET_IMPACT_REPORT_STEPS_TAKEN = gql`
    query impactReportPersonalStepsCompletedCount($id: ID!) {
        community(id: $id) {
            impactReport {
                personalStepsCompletedCount
            }
        }
    }
`;

const GET_STAGES_REPORT_MEMBER_COUNT = gql`
    query communityReportStagesPersonalMemberCount(
        $period: String!
        $id: ID!
        $endDate: ISO8601DateTime!
    ) {
        community(id: $id) {
            report(period: $period, endDate: $endDate) {
                stagesReport {
                    nodes {
                        memberCount
                        stage {
                            name
                        }
                    }
                }
            }
        }
    }
`;

const GET_STAGES_REPORT_STEPS_ADDED = gql`
    query communityReportStagesPersonalStepsAdded(
        $period: String!
        $id: ID!
        $endDate: ISO8601DateTime!
    ) {
        community(id: $id) {
            report(period: $period, endDate: $endDate) {
                stagesReport {
                    nodes {
                        personalStepsAddedCount
                        stage {
                            name
                        }
                    }
                }
            }
        }
    }
`;

const GET_STEPS_COMPLETED_REPORT = gql`
    query communityReportDaysPersonalSteps(
        $period: String!
        $id: ID!
        $endDate: ISO8601DateTime!
    ) {
        community(id: $id) {
            report(period: $period, endDate: $endDate) {
                daysReport {
                    nodes {
                        date
                        personalStepsCompletedCount
                        stageResults {
                            nodes {
                                personalStepsCompletedCount
                                stage {
                                    name
                                }
                            }
                        }
                    }
                }
            }
        }
    }
`;

const GET_TOTAL_STEPS_COMPLETED_SUMMARY = gql`
    query communityReportStagesPersonalStepsCompleted(
        $id: ID!
        $period: String!
        $endDate: ISO8601DateTime!
    ) {
        community(id: $id) {
            report(period: $period, endDate: $endDate) {
                stagesReport {
                    nodes {
                        personalStepsCompletedCount
                        stage {
                            name
                        }
                    }
                }
            }
        }
    }
`;

export {
    GET_IMPACT_REPORT_MOVED,
    GET_IMPACT_REPORT_STEPS_TAKEN,
    GET_STAGES_REPORT_MEMBER_COUNT,
    GET_STEPS_COMPLETED_REPORT,
    GET_STAGES_REPORT_STEPS_ADDED,
    GET_TOTAL_STEPS_COMPLETED_SUMMARY,
};
