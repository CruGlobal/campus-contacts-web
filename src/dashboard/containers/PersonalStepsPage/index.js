import React, { useContext } from 'react';
import { useTranslation } from 'react-i18next';
import moment from 'moment';
import _ from 'lodash';

import Card from '../../components/Card';
import AppContext from '../../appContext';
import StagesSummary from '../../components/StagesSummary';
import StepsChart from '../../components/StepsChart';
import ImpactInfo from '../../components/ImpactInfo';
import FiltersChart from '../../components/FiltersChart';

import {
    GET_IMPACT_REPORT_MOVED,
    GET_IMPACT_REPORT_STEPS_TAKEN,
    GET_STAGES_REPORT_MEMBER_COUNT,
    GET_STEPS_COMPLETED_REPORT,
    GET_STAGES_REPORT_STEPS_ADDED,
    GET_TOTAL_STEPS_COMPLETED_SUMMARY,
} from './queries';

const PersonalStepsPage = () => {
    const { t } = useTranslation('insights');
    const { orgId } = useContext(AppContext);

    return (
        <div>
            <ImpactInfo
                query={GET_IMPACT_REPORT_STEPS_TAKEN}
                text={report =>
                    t('personalSteps.taken', {
                        count: report.impactReport.stepOwnersCount,
                        year: moment().format('YYYY'),
                    })
                }
                variables={{ organizationId: orgId }}
            />
            <Card title={t('personalSteps.completedTotal')}>
                <StagesSummary
                    query={GET_TOTAL_STEPS_COMPLETED_SUMMARY}
                    variables={{
                        organizationId: orgId,
                    }}
                    mapData={data =>
                        data.communityReport.communityStagesReport.map(
                            entry => ({
                                stage: entry.pathwayStage,
                                icon: entry.pathwayStage
                                    .toLowerCase()
                                    .replace(' ', '-'),
                                count: entry.personalStepsCompletedCount,
                            }),
                        )
                    }
                />
            </Card>
            <Card
                title={t('personalSteps.completed')}
                subtitle={t('personalSteps.completedSubtitle')}
            >
                <FiltersChart
                    query={GET_STEPS_COMPLETED_REPORT}
                    mapData={data =>
                        data.communityReport.dayReport.map(row => ({
                            ['total']: row.personalStepsCompletedCount,
                            ['stages']: row.communityStagesReport.map(
                                stage => ({
                                    name: stage.pathwayStage,
                                    count: stage.personalStepsCompletedCount,
                                }),
                            ),
                            ['date']: row.date.substring(0, 2),
                        }))
                    }
                    countAverage={data =>
                        Math.floor(_.sumBy(data, 'total') / data.length)
                    }
                    label={t('personalSteps.legend')}
                />
            </Card>
            <Card
                title={t('personalSteps.added')}
                subtitle={t('personalSteps.addedSubtitle')}
            >
                <StepsChart
                    query={GET_STAGES_REPORT_STEPS_ADDED}
                    mapData={data =>
                        data.organizationPathwayStagesReport.map(row => ({
                            [t('personalSteps.label')]: row.stepsAddedCount,
                            [t('stage')]: row.pathwayStage.name.toUpperCase(),
                        }))
                    }
                    label={t('personalSteps.label')}
                    index={t('stage')}
                    variables={{
                        period: '',
                        organizationId: orgId,
                        endDate: moment().format(),
                    }}
                />
            </Card>
            <ImpactInfo
                query={GET_IMPACT_REPORT_MOVED}
                text={report =>
                    t('personalSteps.reached', {
                        count: report.impactReport.pathwayMovedCount,
                    })
                }
                variables={{ organizationId: orgId }}
            />
            <Card
                title={t('personalSteps.members')}
                subtitle={t('personalSteps.membersSubtitle')}
            >
                <StepsChart
                    query={GET_STAGES_REPORT_MEMBER_COUNT}
                    mapData={data =>
                        data.organizationPathwayStagesReport.map(row => ({
                            [t('members')]: row.memberCount,
                            [t('stage')]: row.pathwayStage.name.toUpperCase(),
                        }))
                    }
                    label={t('members')}
                    index={t('stage')}
                    variables={{
                        period: '',
                        organizationId: orgId,
                        endDate: moment().format(),
                    }}
                />
            </Card>
        </div>
    );
};

export default PersonalStepsPage;

PersonalStepsPage.propTypes = {};
