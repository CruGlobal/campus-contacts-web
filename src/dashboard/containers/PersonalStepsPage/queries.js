import gql from 'graphql-tag';

const GET_IMPACT_REPORT_MOVED = gql`
    query impactReport($organizationId: ID!) {
        impactReport(organizationId: $organizationId) {
            stageProgressionCount
        }
    }
`;

const GET_IMPACT_REPORT_STEPS_TAKEN = gql`
    query impactReport($organizationId: ID!) {
        impactReport(organizationId: $organizationId) {
            personalStepsCount
        }
    }
`;

const GET_STAGES_REPORT_MEMBER_COUNT = gql`
    query organizationStagesReport(
        $period: String!
        $organizationId: ID!
        $endDate: ISO8601DateTime!
    ) {
        organizationStagesReport(
            period: $period
            organizationId: $organizationId
            endDate: $endDate
        ) {
            memberCount
            stage {
                name
            }
        }
    }
`;

const GET_STAGES_REPORT_STEPS_ADDED = gql`
    query organizationStagesReport(
        $period: String!
        $organizationId: ID!
        $endDate: ISO8601DateTime!
    ) {
        organizationStagesReport(
            period: $period
            organizationId: $organizationId
            endDate: $endDate
        ) {
            stage {
                name
            }
            personalStepsAddedCount
        }
    }
`;

const GET_STEPS_COMPLETED_REPORT = gql`
    query communityReport {
        communityReport {
            dayReport {
                date
                personalStepsCompletedCount
                communityStagesReport {
                    pathwayStage
                    personalStepsCompletedCount
                }
            }
        }
    }
`;

const GET_TOTAL_STEPS_COMPLETED_SUMMARY = gql`
    query communityReport {
        communityReport {
            communityStagesReport {
                pathwayStage
                personalStepsCompletedCount
            }
        }
    }
`;

export {
    GET_IMPACT_REPORT_MOVED,
    GET_IMPACT_REPORT_STEPS_TAKEN,
    GET_STAGES_REPORT_MEMBER_COUNT,
    GET_STEPS_COMPLETED_REPORT,
    GET_STAGES_REPORT_STEPS_ADDED,
    GET_TOTAL_STEPS_COMPLETED_SUMMARY,
};
