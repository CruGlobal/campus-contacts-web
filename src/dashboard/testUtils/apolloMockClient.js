import { ApolloClient } from 'apollo-client';
import { InMemoryCache } from 'apollo-cache-inmemory';
import { SchemaLink } from 'apollo-link-schema';
import { addMockFunctionsToSchema } from 'graphql-tools';
import { buildClientSchema } from 'graphql/utilities';
import { createMockerFromIntrospection } from 'fraql/mock';
import gql from 'fraql';

import introspectionQuery from '../../../schema.json';

import { globalMocks } from './globalMocks';

export const createApolloMockClient = (mocks = {}) => {
    const schema = buildClientSchema(introspectionQuery);

    addMockFunctionsToSchema({
        schema,
        mocks: { ...globalMocks, ...mocks },
    });

    return new ApolloClient({
        cache: new InMemoryCache(),
        link: new SchemaLink({ schema }),
    });
};

export const mockFragment = (fragmentDocument, options) => {
    return createMockerFromIntrospection(introspectionQuery, {
        mocks: globalMocks,
    }).mockFragment(gql(fragmentDocument.loc.source.body), options);
};
