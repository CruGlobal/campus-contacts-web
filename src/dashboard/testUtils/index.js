import React, { ReactElement } from 'react';
import { ApolloProvider } from 'react-apollo-hooks';
import { render } from '@testing-library/react';

import { createApolloMockClient } from './apolloMockClient';
import defaultTheme from '../defaultTheme';
import { ThemeProvider } from 'emotion-theming';
import AppContext from '../appContext';

export const renderWithContext = (component, { mocks } = {}, context = {}) => {
    const mockApolloClient = createApolloMockClient(mocks);

    // Warning: don't call any functions in here that return new instances on every call.
    // All the props need to stay the same otherwise renderer won't work.
    const wrapper = ({ children }) => (
        <ApolloProvider client={mockApolloClient}>
            <AppContext.Provider value={context}>
                <ThemeProvider theme={defaultTheme}>{children}</ThemeProvider>
            </AppContext.Provider>
        </ApolloProvider>
    );

    const renderResult = render(component, { wrapper });

    return {
        ...renderResult,
        snapshot: () => {
            expect(renderResult.container).toMatchSnapshot();
        },
    };
};
