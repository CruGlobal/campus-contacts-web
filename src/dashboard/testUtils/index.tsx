import React, { ReactNode } from 'react';
import { ApolloProvider } from 'react-apollo-hooks';
import { render } from '@testing-library/react';
import { ThemeProvider } from 'emotion-theming';

import defaultTheme from '../defaultTheme';
import AppContext from '../appContext';

import { createApolloMockClient } from './apolloMockClient';

export const renderWithContext = (
    component: any,
    { mocks: mocks = {}, appContext = {} } = {},
) => {
    const mockApolloClient = createApolloMockClient(mocks);

    // Warning: don't call any functions in here that return new instances on every call.
    // All the props need to stay the same otherwise renderer won't work.
    const wrapper = ({ children }: { children: ReactNode }) => (
        <ApolloProvider client={mockApolloClient}>
            <AppContext.Provider value={appContext}>
                <ThemeProvider theme={defaultTheme}>{children}</ThemeProvider>
            </AppContext.Provider>
        </ApolloProvider>
    );

    // @ts-ignore
    const renderResult = render(component, { wrapper });
    return {
        ...renderResult,
        snapshot: () => {
            expect(renderResult.container).toMatchSnapshot();
        },
    };
};
