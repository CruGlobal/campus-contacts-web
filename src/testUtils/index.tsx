import React, { ReactNode, ReactElement } from 'react';
import { ApolloProvider } from 'react-apollo-hooks';
import { render } from '@testing-library/react';
import { ThemeProvider } from 'emotion-theming';
import { IMocks } from 'graphql-tools';

import defaultTheme from '../defaultTheme';
import {
    AppContext,
    AppContextValue,
    appContextDefaultValue,
} from '../appContext';

import { createApolloMockClient } from './apolloMockClient';

interface RenderWithContextParams {
    mocks?: IMocks;
    appContext?: AppContextValue;
}

export const renderWithContext = (
    component: ReactElement,
    {
        mocks: mocks = {},
        appContext = appContextDefaultValue,
    }: RenderWithContextParams = {},
) => {
    const mockApolloClient = createApolloMockClient(mocks);

    // Warning: don't call any functions in here that return new instances on every call.
    // All the props need to stay the same otherwise renderer won't work.
    const wrapper = ({ children }: { children?: ReactNode }) => (
        <ApolloProvider client={mockApolloClient}>
            <AppContext.Provider value={appContext}>
                <ThemeProvider theme={defaultTheme}>{children}</ThemeProvider>
            </AppContext.Provider>
        </ApolloProvider>
    );

    const renderResult = render(component, { wrapper });
    return {
        ...renderResult,
        snapshot: () => {
            expect(renderResult.container).toMatchSnapshot();
        },
    };
};
