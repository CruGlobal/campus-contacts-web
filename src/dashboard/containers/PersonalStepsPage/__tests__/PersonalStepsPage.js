import React from 'react';

import { renderWithContext } from '../../../testUtils';
import PersonalStepsPage from '../';

describe('<PersonalStepsPage />', () => {
    it('should render properly', async () => {
        renderWithContext(<PersonalStepsPage />).snapshot();
    });
});
