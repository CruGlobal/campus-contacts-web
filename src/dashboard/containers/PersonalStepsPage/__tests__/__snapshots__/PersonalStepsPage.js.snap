// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PersonalStepsPage /> should render properly 1`] = `
<div>
  <div>
    <h1
      class="css-1c2npbj"
    >
      Loading...
    </h1>
    <div
      class="css-w2m1ij"
    >
      <h1
        class="css-5j0tpf"
      >
        Total Personal Steps of Faith Completed
      </h1>
      <h2
        class="css-140lofm"
      />
      <div
        class="css-1izl14o"
      >
        <div
          class="css-1ohf0ui"
        >
          Loading...
        </div>
      </div>
    </div>
    <div
      class="css-w2m1ij"
    >
      <h1
        class="css-5j0tpf"
      >
        Personal Steps of Faith Completed
      </h1>
      <h2
        class="css-140lofm"
      >
        Personal steps of faith completed by community members.
      </h2>
      <div
        class="css-1izl14o"
      >
        <div>
          Loading...
        </div>
      </div>
    </div>
    <div
      class="css-w2m1ij"
    >
      <h1
        class="css-5j0tpf"
      >
        Personal Steps of Faith Added
      </h1>
      <h2
        class="css-140lofm"
      >
        Current total of personal steps of faith added by members by stage.
      </h2>
      <div
        class="css-1izl14o"
      >
        <div
          class="css-1a3wyyp"
        >
          Loading...
        </div>
      </div>
    </div>
    <h1
      class="css-1c2npbj"
    >
      Loading...
    </h1>
    <div
      class="css-w2m1ij"
    >
      <h1
        class="css-5j0tpf"
      >
        Community Member Stages
      </h1>
      <h2
        class="css-140lofm"
      >
        Current total number of members by stage.
      </h2>
      <div
        class="css-1izl14o"
      >
        <div
          class="css-1a3wyyp"
        >
          Loading...
        </div>
      </div>
    </div>
  </div>
</div>
`;
